### ISSUE TITLE
Annotations on a single panel

### ISSUE THREAD

**Issue by feangulo** [grafana/grafana#717]
Currently we have the option to define annotations that will be applied to all panels in a dashboard. It would be convenient to also have the option to create an annotation that could be applied to only a single panel inside a dashboard.

It could be an additional option when the user enters "Edit" mode in a particular Graphite panel.

**Comment by languitar** [comment]
:+1:

**Comment by faizamodo** [comment]
:+1: our grafana dashboard has a timer panel tracking external calls to another service. Having the ability to track deploy annotations of that other service, on just that panel, would be amazing.

**Comment by Krinkle** [comment]
Disabling annotations per-graph would be quite useful. Especially for dashboards that include a history graph that is e.g. pinned to "Last 6 months" (when the rest of the dashboard shows the last 24 hours or something like that). Otherwise this history graph is obscured with annotations.

**Comment by timgriffiths** [comment]
+1

**Comment by vinhlh** [comment]
:+1:

**Comment by krise3k** [comment]
+1

**Comment by pavleec** [comment]
:+1:

**Comment by sckevmit** [comment]
:+1:

**Comment by addshore** [comment]
+1

**Comment by mtanda** [comment]
+1

**Comment by ygbr** [comment]
+1

**Comment by sslupsky** [comment]
+1

**Comment by rubycut** [comment]
+1

**Comment by mtanda** [comment]
Supporting this is not so difficult.
I already sent PR https://github.com/grafana/grafana/issues/3976. (and close the PR to wait alert feature)
If anyone want to use this ASAP, import above PR and build Grafana locally.

**Comment by GrzegorzWild** [comment]
+1

**Comment by fattylewis** [comment]
+ 1

**Comment by petevk** [comment]
+1

**Comment by n0rad** [comment]
+1

**Comment by andyhollin** [comment]
+1

**Comment by pdf** [comment]
Please don't add `+1` comments, just add a thumbs-up to the original issue description if you wish to express your support, otherwise you just spam everyone who is subscribed to the issue.

**Comment by kylebrandt** [comment]
As noted in #6298 this will also fix the ability to have annotations show up on a panel that has been time-shifted.

**Comment by danievanzyl** [comment]
when will this feature be available?

**Comment by torkelo** [comment]
when it's implemented :) 

So no ETA at the moment :(

**Comment by cherweg** [comment]
+1

**Comment by rpong79** [comment]
+1

**Comment by dkruyt** [comment]
+1

**Comment by LeDistordu** [comment]
+1

**Comment by bgruenefeld** [comment]
+1

**Comment by zicklam** [comment]
+1

**Comment by Wyatts** [comment]
PR #3976 linked above has a comment that this feature needs to wait for alerting. (Why?)  Is alerting in a state where this can be revisited, yet?

It's been over three years since this bug was opened and almost two years since that PR which apparently addressed the main issue.  If it still applies and works, would it be possible to merge it as a stopgap?

**Comment by HerveMARTIN** [comment]
We have the same request here, it seems the good way to use template variable in annotation queries.
In a dashboard approach, dealing with multiple values (concat all values in tags as mentioned by PR #9864) won't match some use cases like dashboard with repeating panel or rows.

**Comment by zer0master** [comment]
I have a strong interest in seeing this realized, as well as being curious now that alerting is in place, why per-panel annotations are not.

Grafana: 4.6.3
InfluxDB: 1.3.6

Here's my use case: as part of a devops group, we maintain multiple virtual environments for development and QA activities. We're making progress with Grafana and InfluxDB in persisting various functional and end-to-end sanity test results to raise visibility on downtime, and ultimately to identify the causes for it.

We need the annotation range that Grafana's native annotations support to visually show duration of downtime to interested groups. Influx appears only capable of point-in-time annotations, which is fine for software drops, but not this particular type of information (in our opinion).

As part of making such updates as easy as possible, we're using a Hubot-based Slack integration to target specific sanity history panels with start/end, essentially a CRUD-based system. We cross-reference user-friendly environment names to the corresponding dashboard/panel IDs to make annotations appear in their associated graph.

What we're finding is that tag queries will let us have differing colors, ideally to distinguish between maintenance and outage windows, but at the expense of appearing on all such panels on the dashboard.

I'm also examining an alternative of embedding the history graphs from their own dashboards in this executive summary, but I seem to lose the per-panel behavior that the dashboard only query provides. We aren't using repeated panels much, though the history graph is currently defined that way; could that be contributing to the problem?

I suppose my questions are as follows:
- have per-panel annotations been tabled indefinitely, and if so, why? based on the number of upvotes here, it seems many users want them, though perhaps for reasons different from my own
- is there currently a way to get this effect of per-panel annotations *and* tag-query specific colors in one dashboard without the "one annotation for all" effect? if not, is there a viable alternative?
- have I overlooked some obvious answer in my attempts? based on the fact that annotation settings appear to be a top-level menu choice, I have assumed to date that I have not

We are thrilled with Grafana for the most part, but this lack of configurabiltiy is becoming a pain point for our organization.

**Comment by jmarceli** [comment]
It seems to be supported but only for annotations created via API. See: http://docs.grafana.org/http_api/annotations/#create-annotation

```
{
  "dashboardId":468,
  "panelId":1,
  "time":1507037197339,
  "isRegion":true,
  "timeEnd":1507180805056,
  "tags":["tag1","tag2"],
  "text":"Annotation Description"
}
```
`dashboardId` and `panelId` params allows to specify exact panel where annotation will appear. It doesn't solve the issue if you want to use annotation query but it might be some kind of workaround.

**Comment by calebtote** [comment]
The above isn't exactly correct -- at least not when using annotation queries. For the example above, if you have an annotation query that looks for `tag1`, the `panelId` stored in the DB is disregarded, and the annotation for `tag1` will be drawn on every panel on the dashboard.

**Comment by gauravk48** [comment]
I have a similar need where I want to invoke a different annotation query for each panel in the dashboard. I don't see any way to do it. Can someone give some pointers?

How can we define multiple queries at the dashboard level and on/off them for each panel? Is there a way to get that working?

Thanks

**Comment by calebtote** [comment]
It's not currently a feature in native Grafana unforunately @gauravk48 -- our team (as in where I work, not Grafana) has an internal feature to add this support. If we finish before Grafana, I'll integrate it -- ETA for us though is still multiple months out.

**Comment by gauravk48** [comment]
Thanks @calebtote for your reply.

**Comment by jwhitcraft** [comment]
@calebtote just wondering if you or your team has made any progress on implementing them.

**Comment by calebtote** [comment]
We got pretty close, but I've since started a new job (and as such, no longer have legal access to the implementation at my old company)

**Comment by inghemagine** [comment]
it would be great to see if this could be looked into again

**Comment by santijopo** [comment]
Has anyone found any workarounds for this?

**Comment by fruch** [comment]
@calebtote can you point to the people that still work at your old company that might still be working on this ?

without this kind of feature the annotation feature is really limited, in my case we have a page with lot of panels, and rendering even 10 annotation  on each one of them takes forever (i.e. more then 3sec each time)

**Comment by fruch** [comment]
I have a suggestion to address this issue in this PR
#16588

comments are welcomed

**Comment by IntegersOfK** [comment]
I have multiple panels, each displaying metrics of a different server. When I make an annotation for an issue on one of those servers, I want to show it on that server's chart only. For that reason, I feel like annotation queries should also work like a regular query at the chart level, or even just allow me to reply with my regular targets _and_ annotations for whatever chart is using that series in the same response as the metrics.

**Comment by podorozhny** [comment]
+1

**Comment by pbaumard** [comment]
Please have a look at the PR #19405 which adds an annotation filter for the graph panel.

![image](https://user-images.githubusercontent.com/6176006/65598996-05b33180-df9d-11e9-8e58-f05c56f93e6e.png)

I hope it can solve use cases described in this ticket.

You can also support my other PR #18812 which adds full interpolation of variables for dashboard annotation tag filter.

**Comment by tjvanderende** [comment]
Why both PR are still pending?

**Comment by cr-ron-balter** [comment]
Any progress on this feature request?
Any help needed?

**Comment by 00pur** [comment]
Hi
A solution is still needed for this and I have a simple idea to make this work.
I've also seen PR 18812, but I'm not sure if it fits all the needs here, since it only works with internal annotations.
There is a good solution to query the MySQL database for annotations with multi-value variables, as described here:
https://grafana.com/docs/grafana/latest/datasources/mysql/#annotations
In the Template function it is possible to define repeated panels with one variable.
I think there is an easy way to filter out annotation records for a panel if there is a field with the same name as the panel repeat variable in the result. 
All records belonging to the panel have the value in this field and match the panel variable.
So, only the field name in the SQL query needs to be configured to have the same name as the panel repeat variable for this great function.
This is the same way as an optional name for the end date/time field.

Example with a template variable named hostname:
SELECT
  UNIX_TIMESTAMP(event_time) AS time,
  event_info AS text,
  CONCAT(loaction, ',', rack) AS tags,
  server_name AS hostname
FROM my_server_annotations
WHERE $__timeFilter(event_time) AND hostname IN([[hostname]])
ORDER BY event_time ASC

Panels that do not have a repeat variable should get all annotations, because there the host in the metric can be used for grouping and there are targets for all hosts in that panel.

**Comment by 00pur** [comment]
SELECT
UNIX_TIMESTAMP(event_time) AS time,
event_info AS text,
CONCAT(loaction, ',', rack) AS tags,
server_name AS hostname
FROM my_server_annotations
WHERE $__timeFilter(event_time) AND **server_name** IN([[hostname]])
ORDER BY event_time ASC

**Comment by qJake** [comment]
So, PR #19405 was closed because of the change to React... and as of May 2021, we still aren't able to hide or filter annotations for specific graphs/panels on a dashboard.

Bummer.

![image](https://user-images.githubusercontent.com/652006/118071073-aae66d80-b36c-11eb-823e-6671393c4413.png)

Any ETA on a React version of this?

**Comment by KRJNK** [comment]
Hello, 

Is this project abandoned? Any chance to get to how is it going?

**Comment by andrepura** [comment]
+1

**Comment by bredmich** [comment]
7 years later and still not resolved?

**Comment by cchpackers1** [comment]
Ridiculous

**Comment by Nowaker** [comment]
TOP12 most upvoted ticket left unresolved for years... https://github.com/grafana/grafana/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by snowwin88** [comment]
8 years passed, still it is not resolved??? sigh....

**Comment by crazyscience** [comment]
bump...Why is this not being handled? It's so simple!

**Comment by kamakazikamikaze** [comment]
Are there any active contributors who are looking to integrate this? Being able to filter annotations per panel based on content or tag(s) with color selection based on specific values would be very helpful as there are many use cases for doing so. Given that there have been three PRs opened and closed for this since 2014 -- two of which were because of upcoming changes to Grafana and not because of the quality of the PR content -- and a number of "bump" comments, this seems to be well-desired by the community.

**Comment by pbaumard** [comment]
We are still using the old graph panel (with #19405 changes) since the new time series panel does not support time regions yet (
#35486). I might work on a new new PR targeted to the time series panel after #35486 is fixed.

**Comment by Mendes11** [comment]
Upvoting this. It's impressive that such an important feature is still missing, even though there's requests from so many people for it since 2014...
